pkg/solana/chainwriter/chain_writer_test.go

```diff
@@ -0,0 +1,702 @@
+		accountLookupConfig := []chainwriter.Lookup{}
+
+		// Fetch derived table map
+		derivedTableMap, staticTableMap, err := cw.ResolveLookupTables(ctx, args, lookupTableConfig)
+		require.NoError(t, err)
+
+		// Resolve account metas
+		accounts, err := chainwriter.GetAddresses(ctx, args, accountLookupConfig, derivedTableMap, rw)
+		require.NoError(t, err)
+
+		// Filter the lookup table addresses based on which accounts are actually used
+		filteredLookupTableMap := cw.FilterLookupTableAddresses(accounts, derivedTableMap, staticTableMap)
+		require.Empty(t, filteredLookupTableMap)
+	})
+
+	t.Run("returns empty map if only constant account lookup required", func(t *testing.T) {
+		accountLookupConfig := []chainwriter.Lookup{
+			chainwriter.AccountConstant{
+				Name:       "Constant",
+				Address:    chainwriter.GetRandomPubKey(t).String(),
+				IsSigner:   false,
+				IsWritable: false,
+			},
+		}
+
+		// Fetch derived table map
+		derivedTableMap, staticTableMap, err := cw.ResolveLookupTables(ctx, args, lookupTableConfig)
+		require.NoError(t, err)
+
+		// Resolve account metas
+		accounts, err := chainwriter.GetAddresses(ctx, args, accountLookupConfig, derivedTableMap, rw)
+		require.NoError(t, err)
+
+		// Filter the lookup table addresses based on which accounts are actually used
+		filteredLookupTableMap := cw.FilterLookupTableAddresses(accounts, derivedTableMap, staticTableMap)
+		require.Empty(t, filteredLookupTableMap)
+	})
+}
+
+func TestChainWriter_SubmitTransaction(t *testing.T) {
+	t.Parallel()
+
+	ctx := tests.Context(t)
+	// mock client
+	rw := clientmocks.NewReaderWriter(t)
+	// mock estimator
+	ge := feemocks.NewEstimator(t)
+	// mock txm
+	txm := txmMocks.NewTxManager(t)
+
+	// setup admin key
+	adminPk, err := solana.NewRandomPrivateKey()
+	require.NoError(t, err)
+	admin := adminPk.PublicKey()
+
+	account1 := chainwriter.GetRandomPubKey(t)
+	account2 := chainwriter.GetRandomPubKey(t)
+
+	seed1 := []byte("seed1")
+	account3 := mustFindPdaProgramAddress(t, [][]byte{seed1}, solana.SystemProgramID)
+
+	// create lookup table addresses
+	seed2 := []byte("seed2")
+	programID := solana.MustPublicKeyFromBase58("6AfuXF6HapDUhQfE4nQG9C1SGtA1YjP3icaJyRfU4RyE")
+	derivedTablePda := mustFindPdaProgramAddress(t, [][]byte{seed2}, programID)
+	// mock data account response from program
+	derivedLookupTablePubkey := mockDataAccountLookupTable(t, rw, derivedTablePda)
+	// mock fetch lookup table addresses call
+	derivedLookupKeys := chainwriter.CreateTestPubKeys(t, 1)
+	mockFetchLookupTableAddresses(t, rw, derivedLookupTablePubkey, derivedLookupKeys)
+
+	// mock static lookup table call
+	staticLookupTablePubkey := chainwriter.GetRandomPubKey(t)
+	staticLookupKeys := chainwriter.CreateTestPubKeys(t, 2)
+	mockFetchLookupTableAddresses(t, rw, staticLookupTablePubkey, staticLookupKeys)
+
+	jsonFile, err := os.Open("testContractIDL.json")
+	require.NoError(t, err)
+
+	defer jsonFile.Close()
+
+	data, err := ioutil.ReadAll(jsonFile)
```

Comment:
nit: [ioutil](https://pkg.go.dev/io/ioutil) package is deprecated. We should prefer [io package](https://pkg.go.dev/io) instead. This will fix the linter issue

quoting from ioutil docs:
> Deprecated: As of Go 1.16, the same functionality is now provided by package [io](https://pkg.go.dev/io) or package [os](https://pkg.go.dev/os), and those implementations should be preferred in new code. See the specific function documentation for details.

quoting from [ioutil.readAll](https://pkg.go.dev/io/ioutil#ReadAll) docs:
> Deprecated: As of Go 1.16, this function simply calls [io.ReadAll](https://pkg.go.dev/io#ReadAll).

```suggestion
	data, err := io.ReadAll(jsonFile)
```